tracer/src/Datadog.Trace/Tagging/TagPropagation.cs

```diff
@@ -0,0 +1,208 @@
+
+        foreach (var headerTag in headerTags)
+        {
+            // the shortest tag has the "_dd.p." prefix, a 1-character key, and 1-character value (e.g. "_dd.p.a=b")
+            if (headerTag.Length >= MinimumPropagationHeaderLength &&
+                headerTag.StartsWith(PropagatedTagPrefix, StringComparison.Ordinal))
+            {
+                // NOTE: the first equals sign is the separator between key/value, but the tag value can contain
+                // additional equals signs, so make sure we only split on the _first_ one. For example,
+                // the "_dd.p.upstream_services" tag will have base64-encoded strings which use '=' for padding.
+                var separatorIndex = headerTag.IndexOf(KeyValueSeparator);
+
+                // "_dd.p.a=b"
+                //         ⬆   separator must be at index 7 or higher and before the end of string
+                //  012345678
+                if (separatorIndex > PropagatedTagPrefixLength &&
+                    separatorIndex < headerTag.Length - 1)
+                {
+                    // TODO: implement something like StringSegment to avoid allocating new (sub)strings?
+                    var key = headerTag.Substring(0, separatorIndex);
+                    var value = headerTag.Substring(separatorIndex + 1);
+
+                    if (IsValid(key, value))
+                    {
+                        traceTags.Add(new(key, value));
+                    }
+                    else
+                    {
+                        // we can't reuse the same string if we skip any key/value pair
+                        cachedHeader = null;
+                    }
+                }
+                else
+                {
+                    // we can't reuse the same string if we skip any key/value pair
+                    cachedHeader = null;
+                }
+            }
+            else
+            {
+                // we can't reuse the same string if we skip any key/value pair
+                cachedHeader = null;
+            }
+        }
+
+        return traceTags.Count > 0 ? new TraceTagCollection(traceTags, cachedHeader) : new TraceTagCollection();
+    }
+
+    /// <summary>
+    /// Constructs a string that can be used for horizontal propagation using the "x-datadog-tags" header
+    /// in a "key1=value1,key2=value2" format. This header should only include tags with the "_dd.p.*" prefix.
+    /// The returned string is cached and reused if no relevant tags are changed between calls.
+    /// </summary>
+    /// <returns>A string that can be used for horizontal propagation using the "x-datadog-tags" header.</returns>
+    public static string ToHeader(TraceTagCollection tags, int maxOutgoingHeaderLength)
+    {
+        if (tags.Count == 0)
+        {
+            return string.Empty;
+        }
+
+        var sb = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
+
+        foreach (var tag in tags)
```

Comment:
In the previous implementation, there was a `lock (_tags)` statement that enclosed this for loop. Is it safe to remove it?

Reply:
I refactored iterating into ToArray() which returns a copy that won't be modified.